x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java

```diff
@@ -95,80 +99,105 @@ private FilterExec rewrite(FilterExec filterExec, EsQueryExec esQueryExec) {
+        // Don't do anything if there are no distances to push down
+        if (distances.isEmpty()) {
+            return filterExec;
+        }
+
+        // Process the EVAL to get all aliases that might be needed in the filter rewrite
+        AttributeMap.Builder<Attribute> aliasReplacedByBuilder = AttributeMap.builder();
+        evalExec.fields().forEach(alias -> {
+            if (alias.child() instanceof Attribute attr) {
+                aliasReplacedByBuilder.put(alias.toAttribute(), attr);
+            }
+        });
+        AttributeMap<Attribute> aliasReplacedBy = aliasReplacedByBuilder.build();
+
+        // First we split the filter into multiple AND'd expressions, and then we evaluate each individually for distance rewrites
+        List<Expression> pushable = new ArrayList<>();
+        List<Expression> nonPushable = new ArrayList<>();
+        for (Expression exp : splitAnd(filterExec.condition())) {
+            Expression resExp = exp.transformUp(ReferenceAttribute.class, r -> aliasReplacedBy.resolve(r, r));
+            // Find and rewrite any binary comparisons that involve a distance function and a literal
+            var rewritten = rewriteDistanceFilters(resExp, distances);
+            // If all pushable StDistance functions were found and re-written, we need to re-write the FILTER/EVAL combination
+            if (rewritten.equals(resExp) == false && conditionHasNoReferences(rewritten, distances)) {
+                pushable.add(rewritten);
+            } else {
+                nonPushable.add(exp);
+            }
+        }
+
+        // If nothing pushable was rewritten, we can return the original filter
+        if (pushable.isEmpty()) {
+            return filterExec;
+        }
+
+        // Move the rewritten pushable filters below the EVAL
+        var distanceFilter = new FilterExec(filterExec.source(), esQueryExec, Predicates.combineAnd(pushable));
+        var newEval = new EvalExec(evalExec.source(), distanceFilter, evalExec.fields());
+        if (nonPushable.isEmpty()) {
+            // No other filters found, we can just return the original eval with the new filter as child
+            return newEval;
+        } else {
+            // Some other filters found, we need to return two filters with the eval in between
+            return new FilterExec(filterExec.source(), newEval, Predicates.combineAnd(nonPushable));
+        }
+    }
+
+    private Map<NameId, StDistance> getPushableDistances(List<Alias> aliases) {
+        Map<NameId, StDistance> distances = new LinkedHashMap<>();
         aliases.forEach(alias -> {
             if (alias.child() instanceof StDistance distance && canPushSpatialFunctionToSource(distance)) {
                 distances.put(alias.id(), distance);
             } else if (alias.child() instanceof ReferenceAttribute ref && distances.containsKey(ref.id())) {
                 StDistance distance = distances.get(ref.id());
                 distances.put(alias.id(), distance);
```

Comment:
Since we rely on `AttributeMap.resolve()` now and that takes renames into account, I think this case is now  obsolete.